### PR TITLE
Feat/licensee select priv purchase

### LINKED ISSUE #209: Licensee selects privileges to buy
dependent on #151 
- OH, KY, NE as test states in all compacts
- states do not automatically get added to production environment
- endpoint recalls all state-related data
- when a user already has a privilege or a home state license in a state, the privilege option is gray

### PR DESCRIPTION
**Pull request by ChiefStief**
### Requirements List
- Make sure your ia-test licensee user will have data returned on the GET `v1/purchases/privileges/options` end point

### Description List
- Implemented Actions, Mutations, network calls, changes to user store and all necessary test changes for new endpoint
- Altered Compact model and tests to support new data
- Created new model, serializer and tests for objects returned in new endpoint
- Implemented privilege purchase screen as shown + military discount
- Implemented mockApi to mock new endpoints and create necessary data scenarios to validate styling
- Added tranlations as needed

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- `yarn test:unit:coverage` should meet our goals
- Code review
- login as licensee user on both the real API and mockAPI and confirm that `:your-compact/Privileges/SelectPrivileges` looks good and behaves well on all screen sizes

Closes #209

### CODE REVIEW

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.vue [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                        :isEnabled="!isFormLoading && isAtLeastOnePrivilegeChosen"
+                    />
+                </div>
+            </div>
+            <div class="select-privileges-core-container">
+                <div class="select-privileges-title">
+                    {{selectPrivilegesTitleText}}
+                </div>
+                <div class="lists-container">
+                    <div class="state-select-list">
```
Comment (jsandoval81) [inline webroot/src/pages/SelectPrivileges/SelectPrivileges.vue:38]: For a11y, lists are ideally semantically tagged, in this case ul / li.

**Review by jsandoval81 — CHANGES_REQUESTED** [review]
Looking pretty crisp so far 👍 

In addition to the comments below, I had a few extra items:

1. I think it makes sense to add the loading indicator in this ticket. It should be pretty straightforward to add the `LoadingSpinner` and just toggle it from the user store `isLoading`. Slack DM me if that's not the case.
2. The confirmation modal looks wider than in the designs, and also frequently has a vertical scroll bar (image below).
    - I'm wondering if a `max-width` is in order, and maybe also an `overflow-y: auto` if needed

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.less [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                    flex-direction: column;
+                    align-items: center;
+
+                    .selected-state-block {
+                        width: 27rem;
+                        margin: 1rem 0;
+                        padding: 1rem 1.6rem;
+                        border-radius: 20px;
+                        color: white;
+                        background-color: @darkBlue;
```
Comment (jsandoval81) [inline webroot/src/pages/SelectPrivileges/SelectPrivileges.less:148]: This might be easier as `@primaryColor` for now (which is black). That's what I've been using in anticipation that primary color will change globally (and hopefully easily) once we implement the hi-fi designs.

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.less [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                    @media @desktopWidth {
+                        max-height: 65vh;
+                    }
+
+                    .state-select-unit {
+                        width: 27rem;
+                        margin: 1rem;
+                        padding: 0 1rem;
+                        border: 1px solid @lightGrey;
+                        border-radius: 10px;
```
Comment (jsandoval81) [inline webroot/src/pages/SelectPrivileges/SelectPrivileges.less:115]: Now that this element is more expectedly interactive, a `cursor: pointer` might be in order.

**Review by jsandoval81 — CHANGES_REQUESTED** [review]
Really nice!

Just a couple very minor style items remaining.

**Review by jsandoval81 — APPROVED** [review]

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.less (hunk 2 above)
Comment (jsandoval81) [reply webroot/src/pages/SelectPrivileges/SelectPrivileges.less]: Resolved.

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.less (hunk 3 above)
Comment (jsandoval81) [reply webroot/src/pages/SelectPrivileges/SelectPrivileges.less:115]: Resolved.

**Comment by jsandoval81** [comment]
@jlkravitz This is ready for your review.

_FYI the last 2 comments wouldn't present us the "mark as resolved" UI. Feel free to mark those as resolved if you have the permission._

**Comment by jlkravitz** [comment]
- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] run a security audit of dependencies (e.g. `npm audit` and `pip audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)

File: webroot/src/models/PrivilegePurchaseOption/PrivilegePurchaseOption.model.spec.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        };
+        i18n.global.locale = 'en';
+    });
+    it('should create a PrivilegePurchaseOption with expected defaults', () => {
+        const privilegePurchaseOption = new PrivilegePurchaseOption();
+
+        // Test field values
+        expect(privilegePurchaseOption).to.be.an.instanceof(PrivilegePurchaseOption);
+        expect(privilegePurchaseOption.jurisdiction).to.be.an.instanceof(State);
```
Comment (jlkravitz) [inline webroot/src/models/PrivilegePurchaseOption/PrivilegePurchaseOption.model.spec.ts:35]: any particular state? should we test that it's an "empty" state?

generally, why would a privilege purchase option ever be instantiated without the requisite variables (jurisdiction, compact, fee, etc)?

File: webroot/src/models/PrivilegePurchaseOption/PrivilegePurchaseOption.model.spec.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        const privilegePurchaseOption = new PrivilegePurchaseOption(data);
+
+        expect(privilegePurchaseOption).to.be.an.instanceof(PrivilegePurchaseOption);
+        expect(privilegePurchaseOption.jurisdiction).to.be.an.instanceof(State);
+        expect(privilegePurchaseOption.jurisdiction.abbrev).to.equal('ca');
+        expect(privilegePurchaseOption.compact).to.equal(data.compact);
+        expect(privilegePurchaseOption.fee).to.equal(5);
+        expect(privilegePurchaseOption.isMilitaryDiscountActive).to.equal(true);
+        expect(privilegePurchaseOption.militaryDiscountType).to.equal(FeeTypes.FLAT_RATE);
+        expect(privilegePurchaseOption.militaryDiscountAmount).to.equal(10);
```
Comment (jlkravitz) [inline webroot/src/models/PrivilegePurchaseOption/PrivilegePurchaseOption.model.spec.ts:62]: note to self: is the calculated final amount tested?

File: webroot/src/network/userApi/interceptors.ts [hunk 6]
```diff
@@ -17,7 +17,7 @@ export const requestSuccess = () => async (requestConfig) => {
     // Endpoints in the below list are accessible to users who are logged in via the licensee userpool
     // and as such require sending that token to authorize their use. All other endpoints are only accessible
     // by users logged in via the staff user pool and will therefor send those tokens to authorize.
```
Comment (jlkravitz) [inline webroot/src/network/userApi/interceptors.ts:19]: just noticed small typo: 
```suggestion
    // by users logged in via the staff user pool and will therefore send those tokens to authorize.
```

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.ts [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+    get privilegeList(): Array<License> {
+        return this.licensee?.privileges || [];
+    }
+
+    get licenseList(): Array<License> {
+        return this.licensee?.licenses || [];
+    }
+
+    get alreadyObtainedPrivilegeStates(): Array<string | null> {
+        return this.licenseList.concat(this.privilegeList).map((license) => license?.issueState?.abbrev || null);
```
Comment (jlkravitz) [inline webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:97]: why not just filter out nulls?

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.ts [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+
+        return date;
+    }
+
+    get subTotalList(): Array<number> {
+        return this.selectedStatePurchaseDataList?.map((purchaseInfo) => {
+            const militaryDiscount = purchaseInfo.isMilitaryDiscountActive && purchaseInfo.militaryDiscountAmount
+                ? purchaseInfo.militaryDiscountAmount : 0;
+
+            return ((purchaseInfo?.fee || 0) + (this.currentCompactCommissionFee || 0) - (militaryDiscount));
```
Comment (jlkravitz) [inline webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:227]: This could in theory be negative, but I assume that's something we will just verify on the configuration end and not worth checking here?

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.ts [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+        };
+
+        this.purchaseDataList?.forEach((purchaseOption) => {
+            const { jurisdiction = new State() } = purchaseOption;
+
+            if (jurisdiction) {
+                const { abbrev } = jurisdiction;
+
+                if (abbrev) {
+                    // const shouldDisable = this.alreadyObtainedPrivilegeStates.includes(abbrev);
```
Comment (jlkravitz) [inline webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:275]: remove comment

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.ts [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+            console.log(this.formData);
+            this.endFormLoading();
+        }
+    }
+
+    handleJurisprudenceClicked(privilegePurchaseInformation) {
+        const stateAbbrev = privilegePurchaseInformation.jurisdiction.abbrev;
+        const newValue = this.formData.jurisprudenceConfirmations[stateAbbrev].value;
+
+        if (newValue === true) {
```
Comment (jlkravitz) [inline webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:304]: I assume `newValue` is always a boolean? I also might rename this to something more specific like `confirmationStatus` (or something else that mirrors your other naming conventions)

```suggestion
        if (newValue) {
```

File: webroot/src/store/user/user.actions.ts [hunk 11]
```diff
@@ -199,4 +199,31 @@ export default {
...
+        commit(MutationTypes.GET_PRIVILEGE_PURCHASE_INFORMATION_REQUEST);
+        return dataApi.getPrivilegePurchaseInformation().then((privilegePurchaseData) => {
+            dispatch('getPrivilegePurchaseInformationSuccess', privilegePurchaseData);
+            return privilegePurchaseData;
+        }).catch((error) => {
+            dispatch('getPrivilegePurchaseInformationFailure', error);
+        });
+    },
+    getPrivilegePurchaseInformationSuccess: ({ dispatch, commit, state }, privilegePurchaseData) => {
+        if (privilegePurchaseData?.compactCommissionFee?.compact === state?.currentCompact?.type) {
```
Comment (jlkravitz) [inline webroot/src/store/user/user.actions.ts:212]: Why would a compact compare to a "type"? If these are both one of `aslp`, etc., perhaps we can rename `compact` to `compactType` for naming consistency?

File: webroot/src/store/user/user.spec.ts [hunk 12]
```diff
@@ -395,4 +398,78 @@ describe('User Store Actions', async () => {
...
+        };
+
+        await actions.getPrivilegePurchaseInformationSuccess({ commit, dispatch, state }, privilegePurchaseData);
+
+        expect(commit.calledOnce).to.equal(true);
+        expect(commit.firstCall.args).to.matchPattern([MutationTypes.GET_PRIVILEGE_PURCHASE_INFORMATION_SUCCESS]);
+        expect(dispatch.calledOnce).to.equal(true);
+        expect([dispatch.firstCall.args[0]]).to.matchPattern(['setCurrentCompact']);
+    });
+    it('should successfully start start get privilege purchase information failure', () => {
```
Comment (jlkravitz) [inline webroot/src/store/user/user.spec.ts:439]:
```suggestion
    it('should successfully start get privilege purchase information failure', () => {
```

**Review by jlkravitz — COMMENTED** [review]
looks good! few questions/comments.

File: webroot/src/models/PrivilegePurchaseOption/PrivilegePurchaseOption.model.spec.ts (hunk 4 above)
Comment (ChiefStief) [reply webroot/src/models/PrivilegePurchaseOption/PrivilegePurchaseOption.model.spec.ts:35]: Generally, I don't think it would be but this test confirms that the defaults are set correctly in case it does.

Ill add a test to confirm the empty state.

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.ts (hunk 7 above)
Comment (ChiefStief) [reply webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:97]: The ts compiler is seeming to struggle to comprehend the type change associated with using .filter so I've done this using a forEach. That is likely why I didn't do this in the first place.

File: webroot/src/pages/SelectPrivileges/SelectPrivileges.ts (hunk 8 above)
Comment (ChiefStief) [reply webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:227]: True; however to your point, I think that would be a bigger issue than displaying a negative number here. I think the front end just reporting what the backend gives it is what we want here. Unless the backend is handling this differently(minimum of zero etc) I dont think we want to display something different.

Comment (ChiefStief) [reply webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:227]: Ok checked with Landon, he said it will throw an error so I think displaying the negative is good in that case because it will tip off that there is something weird and inform on the error to an extent

Comment (jlkravitz) [reply webroot/src/pages/SelectPrivileges/SelectPrivileges.ts:236]: Sounds like a great plan!

**Review by jlkravitz — APPROVED** [review]
@isabeleliassen good to merge!

**Review by isabeleliassen — APPROVED** [review]
